### PR TITLE
[Spec][DFlash] Broadcast rank-0 greedy draft proposal across TP

### PR DESCRIPTION
**Pull request by rchalamala**
## Motivation

Under tensor parallelism, each DFlash draft rank computes its block proposal
as a rank-local argmax over vocab-parallel logits. This covers every greedy
proposal path: the Domino rollout, the plain (no selector) vocab-parallel
head, and the selector lattice when the whole batch is greedy. Ranks that
disagree on the argmax — from reduction-order or kernel-scheduling noise in
the sharded logits — propose different draft tokens for the same request,
and the divergence then propagates into verify inputs and accept state
across the TP group.

The existing TP-sync sites (DFLASH_SELECTOR, DFLASH_ACCEPT_*,
DFLASH_TARGET) cover the other rank-local decisions in the DFlash step;
the greedy *draft* proposal is the one remaining unsynced argmax
decision. This PR adds a sync site for it. Note the sampled selector
block (T>0) is out of scope for this site; it can be covered separately
in a follow-up (DFLASH_DRAFT_SAMPLE).

A mixed selector batch (some rows greedy with `top_k <= 1`, others
sampled) is not all-greedy, so a whole-batch guard would leave its
greedy rows unsynchronized even though each one is still a rank-local
argmax over the selector lattice. The later acceptance broadcasts only
lengths and bonuses, applied to rank-local candidates, so those greedy
rows can still commit different tokens across ranks. The sync therefore
covers greedy rows per-row in mixed batches as well.

## Modifications

- Add `SpecTpSyncSite.DFLASH_DRAFT_GREEDY = 18`. Number 17 is left open
  for the sampled-draft sync (DFLASH_DRAFT_SAMPLE), which is still a
  work in progress.
- Add `DFlashWorkerV2._sync_greedy_draft(draft_next, sampling_info)`: when
  the proposal is a rank-local argmax (Domino, no selector, or all-greedy
  batch), broadcast rank 0's `draft_next` rows through `_tp_sync`. The
  sampled selector block is skipped. In a mixed selector batch, derive
  the per-row greedy mask with `resolve_greedy_mask`, broadcast rank 0's
  full proposal, and take only the greedy rows; sampled rows keep their
  rank-local draws, and a batch with no greedy rows does no collective.
  Where selector sampling is disabled (e.g. NPU), every proposal is a
  rank-local argmax regardless of the requested top_k, so the whole
  proposal tensor is synced like an all-greedy batch.
- Call it once per step right after the draft proposal is computed, before
  the proposal is packed into the draft-token block.
- No default change: `SGLANG_SPEC_TP_SYNC` stays `"all"`, which now
  includes the new site.

## Accuracy Tests

CPU unit tests (run on CPU with
`PYTHONPATH=<worktree>/python`,
`python -m pytest`):

- `test/registered/unit/spec/test_dflash_greedy_draft_tp_sync.py` (new):
  5 tests — greedy selector batch synced; sampled batch without selector
  synced; sampled Domino batch synced; all-sampled selector batch (T>0)
  skipped; mixed selector batch broadcasts rank 0's proposal and copies
  only the greedy rows.
- `test/registered/unit/spec/test_spec_tp_sync_default.py` (new): 4 tests —
  default parses to all sites including DFLASH_DRAFT_GREEDY (number 18,
  slug `dflash-draft-greedy`); default config broadcasts on the greedy
  sites; `"all"` preset re-opts in after negation; default produces no
  narrowed-config warning while `"rng"` does.

Result: 9 passed.

Neighbor check: `test/registered/unit/spec/test_dflash_logits.py` — 8
passed, 2 failed; both failures reproduce identically on the clean base
(aed3fb1cdd) and are environmental (torch inductor cache compiled against
GLIBC_2.38, this host has an older glibc). Not caused by this change.

## Speed Tests and Profiling

No speed impact claimed; not measured (CPU-only validation host).

## Validation limits

- CPU-only unit tests with a fake TP group; no multi-rank GPU broadcast was
  exercised.
- The sync covers greedy proposals only; T>0 sampled rows (including the
  sampled rows of a mixed selector batch) keep rank-local draws until the
  follow-up sampled-draft site (17).
- Inductor-dependent tests in the same directory fail in this environment for
  environmental reasons (glibc), on base and on this branch alike.

## Checklist

- [x] Format your code according to the Format code with pre-commit (ruff check/format clean on changed files).
- [x] Add unit tests according to the Run and add unit tests (see Accuracy Tests).
- [ ] Update documentation (no user-facing docs changes in this PR).
- [ ] Provide accuracy and speed benchmark results (CPU-only validation here; see Validation limits).
- [x] Follow the SGLang code style guidance.

---
### CI States

Latest PR Test (Base): :x: [Run #35420771368](https://github.com/sgl-project/sglang/actions/runs/35420771368)
Latest PR Test (Extra): :x: [Run #35420771127](https://github.com/sgl-project/sglang/actions/runs/35420771127)
Latest PR Test (AMD ROCm 10): :x: [Run #35420771237](https://github.com/sgl-project/sglang/actions/runs/35420771237)

### CODE REVIEW

**Comment by jhcdlg** [comment]
**We measured this exact divergence on real hardware, and applying your diff takes it to zero. Two data points for the Motivation, plus one question about the neighbouring accept site.**

Not a duplicate report. Three things: (1) confirmation that the problem is real and reproducible on multi-rank consumer GPUs; (2) a measurement showing your change closes it on our setup; (3) a question about the accept site, which we are **not** filing as a bug.

### Configuration

- TP=2, `--speculative-algorithm DFLASH`, `--speculative-num-draft-tokens 8`, draft = DFlash2 (compressed-tensors W4A16), target = AWQ 4-bit.
- **No NVLink, no P2P**, two *mismatched* consumer GPUs (one sm_86 20 GB, one sm_89 32 GB), each on PCIe x8, single node, WSL2.
- `temperature=0`  -  fully greedy, no RNG anywhere in the decode loop.
- A per-rank probe logs `commit_lens[:bs]` and `candidates[0, :6]` before `_accept_block` returns; the two ranks are then aligned by step index. The tree is plain upstream code  -  none of our own local patches were present in it.

### What we measured

Four runs on the same machine. Runs C and D are back-to-back, with identical warm-up, identical load command and identical analyzer; **the only difference between them is whether your diff was applied.**

|  | run A | run B | **run C** | **run D (control)** |
|---|---|---|---|---|
| revision | `4cb5aeb` (predates `SpecTpSync`) | `main` `5e9342d` | **`main` + this PR** | **`main` (this PR reverted)** |
| `SGLANG_SPEC_TP_SYNC` |  -  | `all` | `all` | `all` |
| steps compared | 2583 | 3003 | **5000** | **5000** |
| `candidates` differ between ranks | 15 (0.581 %) | 75 (2.498 %) | **0 (0.000 %)** | **105 (2.100 %)** |
| committed length differs between ranks | 2 (0.077 %) | 1 (0.033 %) | **0** | 0 |
| output token differs (candidates equal) | 7 (0.271 %) |  -  | **0** | 25 (0.500 %) |

**Run C: 0 of 5000 steps diverged. Run D: 105 of 5000.** So on our hardware this PR closes the divergence we were seeing.

Two examples from run D, both showing the same shape (the first three drafts agree, the divergence starts a few positions in, at near-tied low token ids):

```
step 70    rank0 commit=[1]  cand=[1710, 351, 19,  8, 310, 220]
           rank1 commit=[1]  cand=[1710, 351, 19, 34,  25, 220]

step 129   rank0 commit=[1]  cand=[96939, 3059, 11, 378,  5, 234]
           rank1 commit=[1]  cand=[96939, 3059, 92,   0, 75, 234]
```

That shape  -  identical prefix, then a split several positions into the draft  -  is what pointed us at the draft proposal rather than at the accept arithmetic.

One more grid from run D that we think is worth having in the record: of the **4895 steps where `candidates` agreed, the committed length still differed in 0 of them**. So on our setup every divergence traces back to the candidate set, not to the accept arithmetic itself. (We reproduced run D's whole grid with a second, independently written log parser as well; it matches the numbers above cell for cell.)

### Why this matters to us

Your Motivation says the existing sites cover the other rank-local decisions in the DFlash step. That matches our scan of the whole file: there are exactly **six `sync()` calls** in `dflash_worker_v2.py`, and `DFLASH_TARGET` is in the default preset, so the "different target tokens feed the next draft round" class is already closed. Run C is consistent with that  -  we could not find a remaining divergence once the draft proposal is synchronised.

If this lands, it likely lets us drop a local end-of-accept-path alignment workaround, which is why we care.

### One question about the accept site (a question, not a claim)

The three accept branches are not symmetric in what they sync:

| line | site | value synced |
|---|---|---|
| 2096 / 2097 | `DFLASH_SELECTOR` | `accept_len`, `bonus` |
| 2109 / 2110 | `DFLASH_ACCEPT_SAMPLE` | `accept_len`, `bonus` |
| 2116 | `DFLASH_ACCEPT_GREEDY` | `target_predict` only |
| 2223 | `DFLASH_TARGET` | `next_token_ids` |

Measured on a clean `main` fetched 2026-09-20: `commit_lens` and `new_seq_lens` never appear as the second argument of any `sync()` in this file (0 occurrences each), while `accept_len` and `bonus` appear twice each. Line numbers move with the tree  -  the stable anchors are the `SpecTpSyncSite` names.

**Question: is the greedy accept decision rank-invariant *by construction*?** For example, is `candidates` produced by a kernel that is already cross-rank consistent, which would make syncing `target_predict` sufficient? If yes, a one-line comment stating that invariant is all we need. If not, syncing `accept_len` + `bonus` in the greedy branch would make the three branches symmetric.

We are **not** asking you to change anything in this PR, and we are not filing this as a bug report. Run C is evidence that we cannot currently produce a divergence that needs it  -  we are asking about the invariant, not reporting a failure.

### Caveats

- All four runs are from **one machine** (no P2P / no NVLink, mixed consumer cards, WSL2). Whether NVLink-connected homogeneous hosts see run D's rate is unknown to us; the mechanism does not look hardware-specific, but the rate probably is.
- The rate moved 0.581 % -> 2.498 % between runs A and B, but source revision, dependency versions and run configuration all changed together, so **do not read that as a regression signal**.
- Run D's coverage is narrower in *requests* than run C's: one of its requests took 308 s instead of the usual ~95 s, so the harness skipped its later rounds. Both runs nevertheless cover the same **5000 aligned steps** under the same analyzer, which is the unit we are comparing.
- `SGLANG_QUANT_PARAM_PATH` was left empty in the test harness (the model class on `main` does not implement `load_kv_cache_scales()`), so FP8 KV scales were neutral; it does not affect the cross-rank comparison being made here.
